### DISCUSSION TITLE
🆕 Public beta: Protected tags

### DISCUSSION THREAD

**Discussion by emmaviolet** [community/community#10906] [Other Feature Feedback, Questions, & Ideas]
## Public beta: Protected tags

👋 We currently have some APIs that will allow repo admins to protect tags on their repos. Details of how this works below!

**Related**
https://github.com/github/feedback/discussions/5529
https://github.community/t/feature-request-protected-tags/1742 
https://github.com/isaacs/github/issues/1091

## What we've built
Our beta tag protection feature gives repo admins the option to protect tags on their repo. If they choose to do so, only maintainers and admins will be able to create these tags, and only admins will be able to modify or delete these tags. Tags are protected by patterns - you could protect all tags by using the “*” pattern, but you don’t have to.

To set up and manage these tag protections, we’ve introduced three endpoints, which any repo admin should be able to use:

**GET /repos/{owner/{repo}/tags/protection**

Returns a list of tag protection rules.

**POST /repos/{owner}/{repo}/tags/protection**

Creates a new tag protection rule. Payload must include a `pattern` - example:

curl -" "Authorization: token $GITHUB_TOK"N" \
   -XPOST -d '{"pattern": "*"}' \
   [https://api.github.com/repos/JasonEtco/testing/tags/protection](https://api.github.com/repos/JasonEtco/testing/tags/protection)

{
  "id": 123456,
  "pattern": "*",
  "created_at": "2022-01-12T12:01:47.094-05:00",
  "updated_at": "2022-01-12T12:01:47.094-05:00"
}

**DELETE /repos/{owner}/{repo}/tags/protection/{tag_protection_id}**

Deletes a tag protection rule.

**Comment by mathomp4** [comment]
Query: What is used for pattern matching? For example, I'd like to protect SemVer tags (vX.Y.Z.T), but I didn't know if the [official SemVer regexes](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string) work with the parser GitHub uses. I could do `v*` but then that means `verycooltag` is protected along with `v2.0.0`.

I mean, I'll be testing it out, but you might reply faster than I'll be able to test. 😄

**Reply by mathomp4** [reply]
Well, I think I answered my own question. It handles neither:
```
❯ gh api /repos/GEOS-ESM/tagprotect/tags/protection -f pattern='^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
{
gh: Bad Request (HTTP 400)
  "message": "Bad Request",
  "documentation_url": "https://docs.github.com/rest/reference/repos#create-tag-protection-state-for-a-repository"
}
❯ gh api /repos/GEOS-ESM/tagprotect/tags/protection -f pattern='^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
gh: Bad Request (HTTP 400)
{
  "message": "Bad Request",
  "documentation_url": "https://docs.github.com/rest/reference/repos#create-tag-protection-state-for-a-repository"
}
```
but does like a simple pattern:
```
❯ gh api /repos/GEOS-ESM/tagprotect/tags/protection -f pattern='v*'
{
  "id": 21,
  "pattern": "v*",
  "created_at": "2022-02-02T13:16:38.824-05:00",
  "updated_at": "2022-02-02T13:16:38.824-05:00"
}
```
Hmm. Okay. maybe time to find a less fancy regex...

**Reply by emmaviolet** [reply]
We're currently using [glob-style pattern matching](https://apidock.com/ruby/File/fnmatch/class) for safety reasons & consistency with branch protection (where we do the same).

**Reply by kaumbakahuma** [reply]
Private is beta and is protected tags.

**Reply by mark-vanderpol-tfs** [reply]
Any more information about the pattern expressions possible to match tags? The glob-style is cumbersome to use to handle say v\d+\.\d+\.\d+(-RC\d+) which is supposed to recognize v2.12.09 and v2.12.10-RC01 (etc)
I am using a few glob rules ( v[0-9].[0-9].* and v[0-9].[0-9][0-9].* ) now to prevent/warn for the creation of tags that start like that, but it's not exhaustive. (Eg would allow v02.09.1-RC-not-valid)

**Comment by ctomc** [comment]
Hey @emmaviolet I would like to try this new feature for "globalid" organization. 
On top of my head I it would also be nice to have an option to extend protected tags to the deployment environments so we could be able to deploy to certain environment only from protected tags (as is currently possible for protected branches)

**Reply by emmaviolet** [reply]
Enabled for you! And thank you for the feedback. Just so I'm clear, how do you currently do this for protected branches? Is that the rule `Require deployments to succeed before merging`?

**Reply by varunkuar8789** [reply]
Hold effective from anywhere

**Comment by brettdh** [comment]
@emmaviolet this is great to see - I'd love to have it enabled for my user, as well as the "theliturgists" organization.

**Reply by kaumbakahuma** [reply]
Yes there's a big great to have done.

**Comment by joaorosado** [comment]
Hi @emmaviolet

Awesome to see this being added.
Long time needed since we are using tags to automatically triggers build releases and there is no easy way to prevent users with Write from triggering them.

Can you please enable it for "OutSystems" organization? (no need for my user)
Thanks

**Comment by vedantmgoyal9** [comment]
I want to get it enabled for me.

**Reply by kaumbakahuma** [reply]
Is ok

**Comment by loris-cr** [comment]
Hi @emmaviolet , if possible please enable this feature for the 'chartrequest' organization, as we're using tags to trigger production builds

**Comment by mathomp4** [comment]
@emmaviolet I can think of a possible use case that might be nice, though it might not be possible to support. It would be nice if certain tag patterns could only be pushed by certain users. In some of my repos, there are semi-official tags that aren't "final" SemVer tags, but are pretty important and we'd prefer not to have those overwritten.

I know I could give that user Maintainer status but with that does come a lot of extra power...

**Reply by emmaviolet** [reply]
We're working on this too @mathomp4! Should be enabled soon

**Reply by joaorosado** [reply]
Nice :)

I was just testing this feature (Thanks for enabling it btw) and also got the same issue on some repositories.
The current scenario is:
 - Team 1 is Admin of the repo, are code owners in the main branch and do releases with v2.* tags.
 - Team 2 has only Write permissions in the repo, are code owners of a release1 branch,  have a branch protection rule to allow them to push to the release1 branch and do releases with v1.* tags
 
I wanted to apply to the tags the same thing, allow Team 2 to be able to create the v1.* tags without making them repository level maintainers.

**Reply by mathomp4** [reply]
@joaorosado Interesting. I hadn't thought of that use case, but that is smart. I suppose I should always think "Team" instead of "Person" in GitHub land as well.

**Comment by galargh** [comment]
Hi @emmaviolet 🎉  It's great news 🚀  I want to use protected tags to protect against accidental Go module releases in orgs I'm a member of. Would you mind enabling it for the following orgs?
- [galorgh](https://github.com/galorgh)
- [protocol](https://github.com/protocol)
- [ipfs](https://github.com/ipfs)
- [ipfs-shipyard](https://github.com/ipfs-shipyard)
- [libp2p](https://github.com/libp2p)
- [ipld](https://github.com/ipld)
- [multiformats](https://github.com/multiformats)
- [filecoin-project](https://github.com/filecoin-project)
- [filecoin-shipyard](https://github.com/filecoin-shipyard)

**Comment by 0xDing** [comment]
Hey @emmaviolet I would like to try this new feature for "@brickdoc" organization.

**Comment by duboisf** [comment]
Could you please enable this feature for the @Flatbook org?

**Comment by emmaviolet** [comment]
Enabled for all requestors now. Thanks, all!

**Reply by vedantmgoyal9** [reply]
How do I check? Please tell 🙏🏻

**Reply by emmaviolet** [reply]
If you try calling those APIs with the PAT token for any repo admin, it should work.

**Reply by vedantmgoyal9** [reply]
Thanks!!!

**Reply by varunkuar8789** [reply]
Give great presentations

**Comment by joshua** [comment]
@emmaviolet Can you please enable for the @acst org as well? Thanks!

**Reply by emmaviolet** [reply]
done!

**Reply by joshua** [reply]
Awesome, Thanks!

**Comment by mdragilev** [comment]
@emmaviolet can you please enable it for [name-redacted] org as well.  Thank you.

**Reply by emmaviolet** [reply]
done!

**Reply by mdragilev** [reply]
Thank you!

**Reply by hello-smile6** [reply]
@mdragilev Your org name is still public, you didn't delete the history.

**Comment by ghost** [comment]
@emmaviolet could you pls enable it for my org @MorphosisApps . Thank you.

**Comment by tedchang77** [comment]
@emmaviolet can you please enable for https://github.com/telus-health. thanks!

**Comment by jeffnappi** [comment]
Hi @emmaviolet - So glad this is happening! Thank you. Please enable for the https://github.com/ClearVoice/ org :)

**Comment by hudymi** [comment]
Does this beta feature works also for private repositories?
@emmaviolet could you enable it for a `travelping` organization? Thank you :)

**Comment by emmaviolet** [comment]
Done for all - and yes @hudymi it works for private repositories.

We've now got this enabled for lots of organizations (which is great!) so I'll stop adding new organizations from here until we release to public beta (hopefully very soon).

I'd love more of your continued feedback, however. Those of you who have tried this - what's working for you and what isn't?

**Reply by hudymi** [reply]
Thank you, as we will play with it, I will update you.

**Reply by hudymi** [reply]
@emmaviolet I have played a little bit with it, and works like a charm :)

What I missing:
- Ability to define users that are allowed to create tags
- Ability to block admins/owners too

For example, I have a list of users responsible for executing a release and don't want to give them admin rights, but they should be able to create tags. Also, the admin/owner not always should have the possibility to create a tag, as may not be aware of the release process, etc. Something like that is supported in branch protection feature.

Anyway, big ➕  for this feature!

**Reply by emmaviolet** [reply]
Thanks for this @hudymi! 

> For example, I have a list of users responsible for executing a release and don't want to give them admin rights, but they should be able to create tags.

Just out of interest, would you also not want to give them maintain rights? Because maintainers can create tags in the current implementation.

Thinking about all the other things you mentioned, and especially being able to block admins/owners too. We haven't yet been able to figure out quite how we'd implement it, but are aware it would be nice to have :-)

**Comment by rsmmr** [comment]
Hi @emmaviolet , I just saw this ... If you don't mind, could please enable this for the `zeek` organization as well? Thanks!

**Comment by varunkuar8789** [comment]
progressively move up based on your needs. This journey is a marathon and not a sprint. Looking forward to serving you inside the members area.

**Comment by Dhruvacube** [comment]
Hey could you enable this feature for @The-4th-Hokage org?

**Comment by irtimir** [comment]
Could you please enable this feature for the @WANNABY org?

**Comment by jrnt30** [comment]
Very great timing, we were just talking about this again recently! 

Would appreciate it if you could add this capability to the @m1finance organization as well please!

**Comment by jimbojsb** [comment]
Hi, we'd love to have this on @ziffmedia, @offers and @ziffdavis orgs

**Comment by Diggsey** [comment]
It would be great for my organization if we could mark specific tags as modifyable only by a specific GitHub App (ie. not even administrators could accidentally modify them). It looks like this would not be supported yet, but it's movement in the right direction 😁

**Comment by emmaviolet** [comment]
👋 Enabled for all. Also, you all should now also be able to set tag protections through the UI under repository repo settings. Would be wonderful if you'd give it a try and let me know if there's any feedback!

If you have an enterprise organization, you can also use [custom repository roles](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-peoples-access-to-your-organization-with-roles/managing-custom-repository-roles-for-an-organization) to grant "create tag" and "delete tag" protections to additional user groups.

**Reply by jimbojsb** [reply]
I have the tag protection UI now but I don't see the ability to create a role that say includes create and delete tag. We are an enterprise customer, and i have the roles UI, but those choices don't appear in the dropdown / autocomplete

**Reply by emmaviolet** [reply]
Thanks so much for this. Was some weird behaviour, but should be fixed now.

**Comment by Dhruvacube** [comment]
Well I have suggesstion, in tag protection , there be should be be a way to allow some specific memebers only to create or edit the tag vai the tag protection rule dashboard.
Yeah I know these can be done via the GitHub Enterprise, but here I am talking for the normal GitHub users or the the GitHub users with pro sub.

**Comment by Nishant1500** [comment]
Sounds interesting, can I get added too? :D

**Comment by jrnt30** [comment]
I believe this is inline with a few other suggestions but my mental model may be a bit different.  

Our organization uses a monorepo for several of our service deployments, where each team is effectively responsible for their service's deployments and promotion.  

Our workflow currently makes use of a service, environment and version specific identifier in the form of a tag to express the desired deployment intent `staging.user-v1.2.0` `production.user-v1.2.0` `production.scheduling-v1.2.0`

It would be very helpful for us to be able to use a CODEOWNERS like structure that would give us the ability of associating a specific tagging pattern to our GitHub group structure.  

Ex: user-team-prod-deployers (a subset of user-team-staging-deployers) being able to push both `production.user-v*` and `staging.user-v*` but NOT another services tag of `production.scheduling-*`

Building up these permissions would still allow "root" privledges of "*" to the monorepo admins

**Comment by faizanakram99** [comment]
When will be public beta available ?

@emmaviolet Would you be kind to enable it for @QbilSoftware meanwhile ? Thank you.

**Comment by ggirotto** [comment]
Hey there. Awesome new beta feature. If possible, we from @olmps would like to try this beta feature as well :)

**Comment by EdwinPhilip** [comment]
Most wanted for my use case. Please enable it for @interviewstreet org

**Comment by ChristophHoyer** [comment]
Hej @emmaviolet!
Could you please activate this beta feature on these two organizations, as well?

https://github.com/orgs/BDC-Test
https://github.com/orgs/bosch-engineering

Thanks in advance.

**Comment by NotMyFault** [comment]
Heyo @emmaviolet ,
could you enable it for @IntellectualSites please?

Thanks in advance, pretty handy feature!

**Comment by bpollman** [comment]
Hey @emmaviolet ,

One more request please 
https://github.com/DeputyApp

Thank you, been awaiting this feature for a long time

**Comment by galah92** [comment]
Would appreciate adding @Vayyar as well.

**Comment by farzaneh53** [comment]
#5217

**Comment by t0rr3sp3dr0** [comment]
Hi @emmaviolet,

Could enable this for @inloco and @0rg4n1z4t10n?

Thanks!

**Comment by ctrombley** [comment]
Hi @emmaviolet, 

Can you please enable this for the @indigo-ag org?

cc @magichair

Thanks!

**Reply by farzaneh53** [reply]
> Hi @emmaviolet,
> 
> Can you please enable this for the @indigo-ag org?
> 
> cc @magichair
> 
> Thanks!

**Comment by twsaari** [comment]
Hi @emmaviolet,

Please enable this for the organization `umich-brcf-bioinf`

Thank you!

**Comment by almir** [comment]
Hi @emmaviolet,

Can you, please, enable this for my user account?

Thanks!

**Comment by mkurz** [comment]
Hi,

can you please enable this feature for the organization https://github.com/playframework? 
Thanks!

**Comment by murray-tait** [comment]
Hi @emmaviolet Can you enable this feature for https://github.com/deathtumble . I'm played with tag based deploy model with the intent to recommend with the consultancy I work in.

**Comment by emmaviolet** [comment]
Hi folks! Just a note to say we've now moved this to public beta so the feature should be available for all public repositories on GitHub and in private repositories on paid plans.

**Reply by mattwynne** [reply]
That's great!

We'd like to use this on https://github.com/cucumber's repos, but we'd want to be able to use it to protect some release environment secrets (i.e. be able to set "Deployment tags" for an Environment as well as "Deployment branches") so as to prevent anyone except people with rights on the protected tag from being able to trigger those release workflows.

Is this possible at the moment? If not do you have plans to add it?

**Reply by emmaviolet** [reply]
Hi @mattwynne! Afraid I think we're a little way away from this. Just so I understand - how are you deploying tags to your environment at the moment? (Just trying to understand whether it's possible to just add a check at the deployment stage)

**Reply by mattwynne** [reply]
Hi Emma, thanks for getting back to me.

So right now, when we want to make a release we push the commit to be released to a `release/*` branch.

We have GitHub Actions workflows (e.g. https://github.com/cucumber/tag-expressions/blob/main/.github/workflows/release-npm.yml) that run off push events to that branch pattern, and pull in secrets from a `Release` Environment. We protect the secrets in that Environment by using branch restrictions on the Environment, so that only the small group of maintainers with rights to push to `release/*` branches could run workflows in that Environment and thereby potentially access those secrets.

At the moment, tagging is done as part of the release (e.g. https://github.com/cucumber/tag-expressions/blob/main/.github/workflows/release-github.yaml) by creating a GitHub Release.

The release branches are not needed after the release has completed (they just contain a commit that's on `main` anyway) so we were thinking that it would be ideal, if we could restrict who could push a `v*` tag, that we could do away with using branches altogether for releases and use a tag as the trigger.

What I'm not clear about though, is how we could still protect the secrets if we weren't using branches.

Does that make sense? Feel free to ask more clarifying questions!

**Comment by radhikari-arch** [comment]
I see that github apps permissions are not working for protected tags. Assigned all repository permissions and even Organization permissions for github app, and still the permission denied. 
```
remote: error: GH006: Protected tag update failed for refs/tags/v6.1.6.        
remote: error: You're not authorized to create a tag    
```
Can anyone suggest if this works with github apps?

**Reply by JasonEtco** [reply]
👋 As users need to be repository admins to bypass a tag protection rule, GitHub Apps also need the `Repository Administration: Write` permission to be able to bypass a tag protection rule. Can you confirm that your app has that permission, and is installed on the repository? Thanks!

**Reply by radhikari-arch** [reply]
The app is on Org level, and I have multiple repos (internal) selectively enabled for this. Also, the app has the permission on Repository Administration: Read/Write.
I even tried with all Read/Write permissions for Repository and even on Organization. We have a CI to auto tagging the repo which uses this Apps for token access, These works fine when we remove the Tag protection (v*) rule. Only when protection rule is added, I am getting the above error.

**Reply by radhikari-arch** [reply]
Confirmed that again from the repo, and here is what I see in the app settings:
![image](https://user-images.githubusercontent.com/70524094/160689649-15b79007-4a67-4189-afe1-f859df6efaa1.png)

**Reply by JasonEtco** [reply]
Hey @radhikari-arch - I tried this out and was able to delete a protected tag from a GitHub App; here's exactly what I did, we can compare notes and see if there's something missing:

* Created a GitHub App with `Repository Administration: Read`, `Repository Contents: Write`.
* Created a tag protection rule with the pattern `v*` on my test repository (and installed the app)
* Manually created a tag, `v1.0.0`
* Via the GitHub App installation, attempted to delete the tag (`DELETE /repos/{owner}/{repo}/git/refs/{ref}` [🔗](https://docs.github.com/en/rest/reference/git#delete-a-reference)) 
  * That failed, return an error like `You're not authorized to delete a protected tag`
* Updated permission to `Repository Administration: Write` and accepted the updated permissions for my installation
* Made the same API call as above and it succeeded, and deleted the tag.

Those steps cover _modifying_ a protected tag. I tried the same thing with the [Create a reference API](https://docs.github.com/en/rest/reference/git#create-a-reference) and got the same results.

**Reply by radhikari-arch** [reply]
@JasonEtco Thanks much. I had a confusion with the token being used during the checkout in my workflow. I was using the built in token of the workflow during checkout so my GitHub App permission was not in effect. With that token explicitly set during checkout action, it is now working for me with those permissions you mentioned for the GitHub App.

**Reply by JasonEtco** [reply]
That's great, glad to hear it. Thanks for all the details @radhikari-arch!

**Reply by kaumbakahuma** [reply]
Good service.
Thanks.

**Reply by joshughes** [reply]
Awesome feature that we just discovered we need. If possible we would love to test it out on our organization github.com/6RiverSystems @emmaviolet

**Reply by emmaviolet** [reply]
This is now shipped publicly on all paid & free public repos.

**Comment by tremblaysimon** [comment]
@emmaviolet Hello, only a simple feedback from my team. Currently the feature is working as expected but I think there is something missing about Deploy Keys with writing access that can't create/delete tags when protected tags are enabled. Currently the error is "protected tag hook declined" - even if the deploy key used has write access. What I understand it should work like branch protection and Deploy keys has the same admin permissions when they have write access. Ref.: https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys

**Reply by twuyts** [reply]
One of the steps in a github action is this:
```
      - env:
          GITHUB_TOKEN: ${{ steps.generate-token.outputs.token }}
        run: |
          tag="${{ steps.bump-semver.outputs.new_version }}"
          git tag "${tag}"
          git push origin "${tag}"
``` 
where `generate-token` generates a PAT for the github-app that has access to this repo and has read/write permission to the contents.

However, the step fails with 
```
remote: error: GH006: Protected tag update failed for refs/tags/v3.0.0.        
remote: error: You're not authorized to create a tag        
To https://github.com/Klarrio/myrepo
 ! [remote rejected] v3.0.0 -> v3.0.0 (protected tag hook declined)
```
The tag protection rule is a simple `v*`, so `v3.0.0` should be allowed.

this looks like the exact problem described by @tremblaysimon. Are we missing a permission somewhere? Or is this a bug?

**Reply by ctoestreich** [reply]
This appears to still be an issue where our action is given admin rights but still can't create a new tag due to tag protection `*` on the repo.

**Reply by SharpEdgeMarshall** [reply]
Workaround https://github.com/orgs/community/discussions/10906#discussioncomment-8717893

**Comment by jonapgartwohat** [comment]
Getting an error from an action (https://github.com/mathieudutour/github-tag-action) when it attempts to write a protected tag. 

`Error: Reference update failed`

When I removed tag protection the action succeeded.

**Reply by fractiunate** [reply]
This! saved me so much time. We should reach out to maintainer of (https://github.com/mathieudutour/github-tag-action)

**Reply by tremblaysimon** [reply]
I think it's a similar problem that I described here on the same discussion: https://github.com/orgs/community/discussions/10906#discussioncomment-2633741

Currently it's not possible by a CI system to create a new tag if there is a policy that doesn't allow it. It's not possible to bypass that policy like protected branches allow.

**Comment by cyrilico** [comment]
Hi, any plans to make this configurable at the organization level as well, so that it can be enforced for all repositories in it?

**Comment by cardonator** [comment]
It seems like this feature launched without the ability for the GHA actor to push to protected tags. I want to prevent normal users of the repo from pushing tags in a certain format because it screws up our tag and release job, but I can't have both a tag protection rule and run my tag and release job using the actions actor so I can't really use this feature.

**Reply by tremblaysimon** [reply]
Same problem as discussed here I think. https://github.com/orgs/community/discussions/10906#discussioncomment-2633741

@emmaviolet do you have any updates about that?

**Reply by emmaviolet** [reply]
No, but looking into it!

**Reply by cardonator** [reply]
Thanks for pointing that out @tremblaysimon . One note is that it would be helpful for the default GITHUB_TOKEN (accessed via `GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}`) to be able to have permission to write protected tags.

**Reply by weichunnn** [reply]
@emmaviolet wanted to bump this up to know if there are any updates on this issue :)

**Reply by s0undt3ch** [reply]
Sorry, another bump as our team just faced this issue recently.

**Reply by SharpEdgeMarshall** [reply]
Another bump

**Reply by patrick-knight** [reply]
There are no planned enhancements to protected tags. [Repository tag rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets#about-rulesets-protected-branches-and-protected-tags) are the where future work will happen in this space.  Specifically, on allowing GHA to bypass rules there are several concerns we have today for allowing that from a security standpoint. Rules do allow bypass via Apps, in addition to roles and teams, which would be the current recommendation.

**Reply by SharpEdgeMarshall** [reply]
Workaround https://github.com/orgs/community/discussions/10906#discussioncomment-8717893

**Comment by sammcj** [comment]
As per https://github.com/orgs/community/discussions/31177 being able to protect tags from deletion but not creation and/or the ability to protect a Regex/glob of tags (e.g. v1.0.0 but not v1.0.0-alpha) is really critical for us.

**Reply by emmaviolet** [reply]
This should be possible - responded on the linked discussion

**Reply by L1ghtman2k** [reply]
The `e.g. v1.0.0 but not v1.0.0-alpha` part is not possible AFAIK because full regex is not possible

**Comment by datosh** [comment]
I use `github.ref_protected` in my GitHub actions and expected it to evaluate to `true` when triggered on a protected tag. Sadly, this was not the case. 
Might this be a bug? Is there another conditional I can check to see if an action is running on a protected tag?

**Reply by gladkikhartem** [reply]
@datosh I have exactly the same issue. Probably a bug

**Reply by datosh** [reply]
Thanks for letting me know @gladkikhartem . Hard to tell if it's a bug since the docs are not explicit about expected behavior.

**Reply by datosh** [reply]
Could you have a look @emmaviolet , whether this is expected or a bug?

**Reply by computnik** [reply]
seems like a bug to me too! It limits to using only branches for Environment protection rules for actions!
@emmaviolet thoughts?

**Reply by rcwbr** [reply]
Seems to still be an issue -- more details [here](https://github.com/orgs/community/discussions/142985)

**Comment by tleerai** [comment]
So, am I understanding correctly, there is no way to prevent someone from pushing a tag from an arbitrary branch?

So hypothetically speaking, if I have CICD that fires on every tag push, I can't stop people from making up random branch names and pushing tags?

Or if I just don't want tags to be in the repo for any other reason except when an allowed person creates one?

**Comment by emmaviolet** [comment]
Apologies, folks, I have been reading along here but haven't managed to get this prioritised yet. Sending over to another team who may be able to assist.

**Comment by L1ghtman2k** [comment]
For our use case, we would like to require our protected tags for production to be based on protected branches. This would leverage a lot of features offered by protected branches (status checks, reviewers, PRs, etc)

**Comment by L1ghtman2k** [comment]
It would be also nice to have a dedicated event, that apps can subscribe to, like https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads#branch_protection_rule. 

For our use case we would like our github app to update AWS Assume policy on a role (OIDC), based on protected branches set in the repository.

**Comment by patrick-knight** [comment]
Hey y'all,

We've introduced the beta of [repository rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets#about-rulesets-protected-branches-and-protected-tags), which include Tag rulesets, which include _most_ of the feature you loved from branch protections and brought them to tags.  

If you're on GitHub enterprise you can also protect tag from the org level for all or some repos. Check it out and let us know what you think.

**Comment by dragonx** [comment]
I'm not sure if I'm misunderstanding how this feature works or if this is a bug, but it's not working the way I thought it would work after reading the docs.

I've created a protected tag called `portal-v*`, which I thought would prevent users with "Write" access to the repo from creating tags like `portal-v5.0.0-202309060000`, which we use to kick off a release on our CI server.  I expected only users with "Admin" or "Maintain" permissions to be able to create the tag.

However, a user on our team with "Write" permission was able to create a Release with a name similar to that above, and was able to kick off a release.  Am I misunderstanding this feature?

**Reply by patrick-knight** [reply]
Hard to say exactly what's occurring with out more details, so might be worth opening a support case. But what you are describing is how protected tags function.

It might also worth testing out [rulesets for protected tags ](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository)as that feature is now GA.

**Comment by gdamjan-loka** [comment]
Is it possible to make a ruleset that only allows tags to be made on commits that are on the `main` branch?

**Reply by patrick-knight** [reply]
That is currently not possible. So you would be looking for a way to limit tag creation to only commits that must be reachable from `main` and no other branches? Just checking as they are typically used to notate specific commits as the branch can be an ephemeral ref.

**Comment by GMZwinge** [comment]
There may be a permission for protected tag missing from [permissions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions) to allow the `GITHUB_TOKEN ` to push protected tag. I tried `permissions: write-all` and the `git push --tags` still failed with this error:
```
remote: error: GH006: Protected tag update failed for refs/tags/v<version>.
remote: error: You're not authorized to create a tag        
To https://github.com/<user>/<repo>
 ! [remote rejected] v<version> -> v<version> (protected tag hook declined)
error: failed to push some refs to 'https://github.com/<user>/<repo>'
```

**Reply by patrick-knight** [reply]
Correct, and there are no plans at the moment to allow that. A similar limitation exist for actions and bypassing branch protection rules.

The best way to accomplish this today would be with a GitHub App. [GitHub Apps require the Repository administration: write permission to modify a protected tag.](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules#about-tag-protection-rules)

**Reply by SharpEdgeMarshall** [reply]
Workaround https://github.com/orgs/community/discussions/10906#discussioncomment-8717893

**Reply by huxuan** [reply]
> [GitHub Apps require the Repository administration: write permission to modify a protected tag.](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules#about-tag-protection-rules)

Hi @patrick-knight , I tried a GitHub App with the corresponding permission but it does not seem to work. Any ideas to debug?

![image](https://github.com/community/community/assets/726061/c22880ae-98e0-42f4-8ef7-0a9a30d59bef)

**Reply by huxuan** [reply]
It turns out to be a issue with actions/checkout, in short, we should set the `persist-credentials: false`. For more details, please refer to https://github.com/actions/create-github-app-token/discussions/75#discussioncomment-7469758

**Comment by SharpEdgeMarshall** [comment]
Found a solution to the problem of pushing tags from the CI that involves the usage of a write deploy key and is safe you add the key as a secret to an environment bound to a branch/tag:
```
      - uses: actions/checkout@v4
        with:
          ssh-key: ${{ secrets.SSH_KEY }} # This is you deploy key 
      - name: Commit Bump
        run: |
          MESSAGE="Release <MY_TAG>"
          git config user.name "${GITHUB_ACTOR}"
          git config user.email "${GITHUB_ACTOR}@users.noreply.github.com"
          git tag -fa "<MY_TAG>" -m "<MY_MESSAGE>"
          git push --force origin "<MY_TAG>"
```
